packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java

```diff
@@ -24,14 +30,108 @@
 
 public class FlutterWebView implements PlatformView, MethodCallHandler {
   private static final String JS_CHANNEL_NAMES_FIELD = "javascriptChannelNames";
-  private final WebView webView;
+  private final InputAwareWebView webView;
   private final MethodChannel methodChannel;
   private final FlutterWebViewClient flutterWebViewClient;
   private final Handler platformThreadHandler;
+  private final View flutterView;
+
+  private static class InputAwareWebView extends WebView {
+    private static final String TAG = "InputAwareWebView";
+    private final View flutterView;
+
+    InputAwareWebView(Context context, View flutterView) {
+      super(context);
+      this.flutterView = flutterView;
+    }
+
+    private View threadedInputConnectionProxyView;
+
+    private ThreadedInputConnectionProxyAdapterView proxyAdapterView;
+
+    @Override
+    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
+      if (proxyAdapterView == null) {
+        // No proxy adapter set, so we use the default implementation.
+        return super.onCreateInputConnection(outAttrs);
+      }
+
+      if (!proxyAdapterView.isTriggerDelayed()) {
+        // Currently on the IME thread. Delegate to the superclass.
+        return super.onCreateInputConnection(outAttrs);
+      }
+
+      InputConnection result = super.onCreateInputConnection(outAttrs);
+      if (result != null) {
+        // This should never happen.
+        Log.wtf(TAG, "Failed unexpectedly creating a webview input connection.");
+      }
+
+      final InputMethodManager imm =
+          (InputMethodManager) getContext().getSystemService(INPUT_METHOD_SERVICE);
+      proxyAdapterView.requestFocus();
+      final View containerView = this;
+
+      // This is the crucial trick that gets the InputConnection creation to happen on the correct
+      // thread. https://cs.chromium.org/chromium/src/content/public/android/java/src/org/chromium/content/browser/input/ThreadedInputConnectionFactory.java?l=169.
+      this.post(
+          new Runnable() {
+            @Override
+            public void run() {
+              // This is a hack to make InputMethodManager believe that the proxy view now has a focus.
```

Comment:
I feel like we need to explain that WebView is pulling that trick, which breaks our input connection proxying, so we are pulling the same trick to make it work. If we're moving all of this to a new file it will probably fit as a general explanation there.